### PR TITLE
Add timeDerivative method to Function

### PR DESCRIPTION
**Pull request by bwspenc**
#4769

### CODE REVIEW

File: framework/include/functions/Function.h [hunk 1]
```diff
@@ -87,6 +87,20 @@ class Function :
...
+   * \param p The point in space (x,y,z)
+   * \return The time derivative of the function at the specified time and location
+   */
+  virtual Real timeDerivative(Real t, const Point & p);
+
+  /**
+    * Check to see whether function can calculate time derivative
+    * \return true if time derivative is defined, false otherwise
+    */
+  virtual inline bool haveTimeDerivative() { return false; }
```
Comment (dschwen) [inline framework/include/functions/Function.h]: There is somewhat inconsistent use of ```have``` and ```has``` for such functions in MOOSE. 

Guys! Let's start a five page thread about that!!!!

**Comment by dschwen** [comment]
:+1: looks good to me. Thanks for taking care of ```MooseParsedFunction``` as well!

**Comment by aeslaughter** [comment]
While we are arguing we should look at this: http://www.gockelhut.com/c++/articles/has_member

**Comment by dschwen** [comment]
Uhm, is that a SFINAE based approach? Wait a minute, isn't that for compile time checks?

**Comment by aeslaughter** [comment]
I didn't look that close, I just stirring the pot a bit.

**Comment by bwspenc** [comment]
I just updated the commit to remove the ```hasTimeDerivative()``` methods.  This still keeps the mooseError in ```Function::timeDerivative()``` per our discussion.

**Comment by YaqiWang** [comment]
@bwspenc you may want to add mooseError in gradient() as well. Few source files will need to be fixed under moose/test if you do so.

**Comment by bwspenc** [comment]
@YaqiWang I think that's a good idea to add mooseError to gradient() for consistency with the other methods in the base class, but that would likely have more far-reaching consequences out in the apps, so I'd rather keep that separate from this. This is focused on my immediate need to get the time derivative of a few different types of functions.

**Comment by permcody** [comment]
Agreed, we'll handle that in another PR

**Pull request merged**